crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -394,6 +394,29 @@ impl<'db> TypeInferenceBuilder<'db> {
                 .expressions
                 .extend(deferred_expression_types.iter());
         }
+
+        let class_definitions: Vec<ClassType<'db>> = self
+            .types
+            .declarations
+            .values()
+            .filter_map(|ty| ty.into_class_type())
+            .collect();
+
+        for class in class_definitions {
+            if let Some(mro_errors) = class.mro_possibilities(self.db).possible_errors() {
+                for error in mro_errors {
+                    self.add_diagnostic(
+                        class.node(self.db).into(),
+                        "inconsistent-mro",
+                        format_args!(
+                            "Cannot create a consistent method resolution order (MRO) for class `{}` with possible bases list `[{}]`",
+                            class.name(self.db),
+                            error.iter().map(|base|base.display(self.db)).join(", ")
+                        )
+                    );
+                }
+            }
+        }
```

Comment:
It might help to move out of `declarations` for better performance. 

```suggestion
        let declarations = std::mem::take(&mut self.types.declarations);

        let class_definitions = declarations.values().filter_map(|ty| ty.into_class_type());

        for class in class_definitions {
            if let Some(mro_errors) = class.mro_possibilities(self.db).possible_errors() {
                for error in mro_errors {
                    self.add_diagnostic(
                        class.node(self.db).into(),
                        "inconsistent-mro",
                        format_args!(
                            "Cannot create a consistent method resolution order (MRO) for class `{}` with possible bases list `[{}]`",
                            class.name(self.db),
                            error.iter().map(|base|base.display(self.db)).join(", ")
                        )
                    );
                }
            }
        }

        self.types.declarations = declarations;
```